lib/fizzy/parser_expr.cpp

```diff
@@ -136,9 +136,16 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end, const Mod
         auto& frame = control_stack.top();
         const auto& metrics = metrics_table[opcode];
 
-        if (!frame.unreachable &&
-            (frame.stack_height - frame.parent_stack_height) < metrics.stack_height_required)
-            throw validation_error{"stack underflow"};
+        if (!frame.unreachable)
+        {
+            if ((frame.stack_height - frame.parent_stack_height) < metrics.stack_height_required)
+                throw validation_error{"stack underflow"};
+
+            // Update code's max_stack_height using frame.stack_height of the previous instruction.
+            // The frame.stack_height may have been updated by call instruction and it is fine
+            // to omit value for end instruction.
```

Comment:
I'm not sure what do you mean by "it is fine to omit value for end instruction."? Is it about final end instruction?

Reply:
The each `end` ending a frame. I will update the comment to

> The frame.stack_height may have been updated by call instruction. This also omits every `end` and `else` instructions as they pop/reset the frame object.

Reply:
Still not sure I understand the second part. Do you mean that there's no need to update `max_stack_height` on `end` and `else`? But it's kinda not related to why we update `max_stack_height` according to previous instruction...

I would say we don't update `max_stack_height` on `end` and `else`, because `frame.stack_heigh` doesn't grow there anyway. Maybe here I'd just remove the second sentence.

But also wouldn't it be more straightforward to update here after `frame.stack_height` growth (for current instruction) + in one other place - `update_caller_frame`. No need for complicated comment here then. But you would need to pass `code` there...

Reply:
I try to explain why the last update `frame.stack_height += metrics.stack_height_change` is ignored for `code.max_stack_height`.

I was considering changing `update_caller_frame` that way, but decided to try having the update in single place (and this is result of it).

So removing last sentence is good enough resolution? Or maybe move it to `frame.stack_height += metrics.stack_height_change`?

Reply:
If we keep the update here, I'd change the comment to something like
```
// Update code's max_stack_height using frame.stack_height of the previous instruction.
// At this point `frame.stack_height` includes additional changes to the stack if previous instruction was a call.
// This way this update is skipped for the end/else instruction (becaue their frame is already popped/reset), but it doesn't matter because they don't grow stack anyway.
```

Reply:
Comment updated.